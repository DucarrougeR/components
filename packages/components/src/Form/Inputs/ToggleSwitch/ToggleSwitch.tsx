import { rem, rgba } from 'polished'
import React, { FunctionComponent } from 'react'
import styled, { css, StyledComponent } from 'styled-components'
import { palette } from '@looker/design-tokens'
import { Omit, ThemedProps } from '../../../../types'
import { CustomizableAttributes } from '../../../../types/attributes'
import { Box, BoxProps } from '../../../Box'
import { Checkbox } from '../Checkbox'

// Omit<T, K> is built in to TypeScript 3.5, delete next line when we upgrade
type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export interface KnobProps {
  className?: string
  size: number
  disabled?: boolean
  on?: boolean
}

export interface ToggleSwitchProps
  extends BoxProps<HTMLInputElement>,
    Omit<KnobProps, 'size'> {
  size?: number
}

type KnobComponentType = FunctionComponent<KnobProps>
type StyledKnobComponentType = StyledComponent<KnobComponentType, KnobProps>

const knobTransform = (props: ThemedProps<KnobProps>) => {
  const transform = props.on ? `translateX(${rem(props.size * 0.75)})` : ''
  return css`
    transform: ${transform};
    transition: ${props.theme.transitions.durationModerate};
  `
}

const KnobFactory: React.FC<KnobProps> = ({ className, ...props }) => (
  <Box
    className={className}
    position="absolute"
    bottom={rem(props.size * 0.1)}
    left={rem(props.size * 0.1)}
    width={rem(props.size * 0.8)}
    height={rem(props.size * 0.8)}
    borderRadius="50%"
    bg={
      props.on
        ? CustomizableToggleSwitchAttributes.knobOnColor
        : CustomizableToggleSwitchAttributes.knobOffColor
    }
  />
)

const Knob: StyledKnobComponentType = styled<KnobComponentType>(KnobFactory)`
  ${knobTransform};
`

const KnobContainerFactory: KnobComponentType = ({ className, ...props }) => {
  const hoverStyle = props.disabled
    ? undefined
    : { boxShadow: `0 0 .01rem 0.01rem ${rgba(palette.primary500, 0.5)}` }
  return (
    <Box
      className={className}
      position="absolute"
      top="0"
      bottom="0"
      left="0"
      right="0"
      borderRadius={rem(props.size)}
      bg={
        props.on
          ? CustomizableToggleSwitchAttributes.onColor
          : CustomizableToggleSwitchAttributes.offColor
      }
      hoverStyle={hoverStyle}
    >
      <Knob {...props} />
    </Box>
  )
}

const KnobContainer: StyledKnobComponentType = styled<KnobComponentType>(
  KnobContainerFactory
)`
  transition: ${props => props.theme.transitions.durationModerate};
`

export type ToggleSwitchComponentType = FunctionComponent<ToggleSwitchProps>
export type StyledToggleSwitchComponentType = StyledComponent<
  ToggleSwitchComponentType,
  ToggleSwitchProps
>

const InternalToggleSwitch: ToggleSwitchComponentType = ({
  size = 20,
  disabled,
  on,
  ...props
}) => {
  return (
    <Box
      width={rem(size * 1.75)}
      height={rem(size)}
      display="inline-block"
      position="relative"
      verticalAlign="middle"
      cursor={!disabled ? 'pointer' : undefined}
    >
      <Checkbox
        checked={on}
        disabled={disabled}
        opacity={0}
        width="100%"
        height="100%"
        role="switch"
        aria-checked={on}
        position="absolute"
        top="0"
        left="0"
        zIndex={1}
        {...props}
      />
      <KnobContainer size={size} on={on} disabled={disabled} />
      {disabled && (
        <Box
          position="absolute"
          top="0"
          bottom="0"
          left="0"
          right="0"
          opacity={0.4}
          bg={palette.charcoal300}
          borderRadius={rem(size)}
        />
      )}
    </Box>
  )
}

/** @component */
export const ToggleSwitch: StyledToggleSwitchComponentType = styled<
  ToggleSwitchComponentType
>(InternalToggleSwitch)`
  :focus + ${Box} {
    box-shadow: 0 0 0 0.2rem ${rgba(palette.primary500, 0.4)};
  }
`

export interface CustomizableToggleSwitchAttributes
  extends CustomizableAttributes {
  knobOnColor: string
  knobOffColor: string
  onColor: string
  offColor: string
}

export const CustomizableToggleSwitchAttributes: CustomizableToggleSwitchAttributes = {
  knobOffColor: palette.white,
  knobOnColor: palette.white,
  offColor: palette.charcoal300,
  onColor: palette.purple300,
}
