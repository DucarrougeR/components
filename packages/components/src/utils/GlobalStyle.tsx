import { css, createGlobalStyle } from 'styled-components'

/*
 * @TODO - Explore local loading of font files.
 *

import openSansSemi from '../fonts/open-sans-600.woff2'
import openSansRegular from '../fonts/open-sans-regular.woff2'

const fontFaces = css`
  @font-face {
    font-family: 'Open Sans';
    font-style: normal;
    font-weight: 400;
    src: local('Open Sans Regular'), local('OpenSans-Regular'),
      url(${openSansRegular}) format('woff');
  }

  @font-face {
    font-family: 'Open Sans';
    font-style: normal;
    font-weight: 600;
    src: local('Open Sans SemiBold'), local('OpenSans-SemiBold'),
      url(${openSansSemi}) format('woff');
  }
`
*/

const fonts = css`
  @import url('https://fonts.googleapis.com/css?family=Open+Sans:400,600&display=swap');

  body,
  button,
  input,
  textarea,
  select {
    font-family: 'Open Sans', sans-serif;
  }
`

const reset = css`
  * {
    padding: 0;
    box-sizing: border-box;
  }

  ul,
  li,
  p,
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    margin: 0;
  }

  a {
    text-decoration: none;
  }

  html,
  body {
    width: 100%;
    height: 100%;
    margin: 0;
  }

  body {
    background-color: ${props => props.theme.colors.palette.white};
  }
`

export const GlobalStyle = createGlobalStyle`
  ${fonts}
  ${reset}
`
