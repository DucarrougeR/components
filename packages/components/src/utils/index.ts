export * from './GlobalStyle'
export * from './useToggle'
