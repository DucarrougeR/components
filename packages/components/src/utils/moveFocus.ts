/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

export const getTabStops = (ref: HTMLElement): HTMLElement[] =>
  Array.from(
    ref.querySelectorAll(
      'a,button:not(:disabled),[tabindex="0"],[tabindex="-1"]'
    )
  )

export const moveFocus = (
  direction: number,
  initial: number,
  element?: HTMLElement | null
) => {
  if (element) {
    const tabStops = getTabStops(element)

    if (tabStops.length > 0) {
      let toFocus = tabStops.slice(initial)[0]
      if (
        document.activeElement &&
        tabStops.includes(document.activeElement as HTMLElement)
      ) {
        const next =
          tabStops.findIndex((f) => f === document.activeElement) + direction

        if (next === tabStops.length) return null
        if (!tabStops[next]) return null
        toFocus = tabStops[next]
      }
      toFocus.focus()
      return toFocus
    }
  }
  return null
}
