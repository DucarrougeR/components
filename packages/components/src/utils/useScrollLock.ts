import { RefObject, useEffect } from 'react'

export function useScrollLock(
  disabled?: boolean,
  allowScrollWithin?: RefObject<HTMLElement | null>
) {
  useEffect(() => {
    let scrollTop = window.scrollY
    let scrollTarget: EventTarget | HTMLElement | null = document

    function stopScroll(e: Event) {
      if (e.target !== null && e.target !== scrollTarget) {
        scrollTarget = e.target
        scrollTop =
          scrollTarget instanceof Element
            ? scrollTarget.scrollTop
            : window.scrollY
      }

      if (
        scrollTarget instanceof Element &&
        !(
          allowScrollWithin &&
          allowScrollWithin.current &&
          allowScrollWithin.current.contains(scrollTarget)
        )
      ) {
        scrollTarget.scrollTop = scrollTop
      } else if (scrollTarget === document) {
        window.scrollTo({ top: scrollTop })
      }
    }

    if (disabled) {
      window.removeEventListener('scroll', stopScroll, true)
    } else {
      window.addEventListener('scroll', stopScroll, true)
    }

    return () => {
      window.removeEventListener('scroll', stopScroll, true)
    }
  }, [disabled, allowScrollWithin])
}
