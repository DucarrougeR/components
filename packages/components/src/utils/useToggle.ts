import { useState } from 'react'

export interface UseToggleReturn {
  value: boolean
  setOn: () => void
  setOff: () => void
  toggle: () => void
}

export function useToggle(initialValue = false): UseToggleReturn {
  const [value, setValue] = useState(initialValue)
  const setOn = () => setValue(true)
  const setOff = () => setValue(false)
  const toggle = () => setValue(!value)
  return { value, setOn, setOff, toggle }
}
