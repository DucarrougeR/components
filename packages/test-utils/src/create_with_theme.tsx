import { theme } from '@looker/design-tokens'
import { render } from '@testing-library/react'
import { mount, shallow } from 'enzyme'
import 'jest-styled-components'
import * as React from 'react'
import { create } from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'

export const withThemeProvider = (Component: React.ReactElement<any>) => (
  <ThemeProvider theme={theme}>{Component}</ThemeProvider>
)

export const createWithTheme = (Component: React.ReactElement<any>) =>
  create(withThemeProvider(Component))

export const mountWithTheme = (Component: React.ReactElement<any>) =>
  mount(withThemeProvider(Component))

export const shallowWithTheme = (Component: React.ReactElement<any>) =>
  shallow(withThemeProvider(Component))

export const renderWithTheme = (Component: React.ReactElement<any>) =>
  render(withThemeProvider(Component))
