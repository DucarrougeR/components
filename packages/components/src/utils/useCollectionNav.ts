/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { FocusEvent, KeyboardEvent, useState } from 'react'
import { getNextFocusTarget } from './getNextFocusTarget'
import { getTabStops, moveFocus } from './moveFocus'
import { useWrapEvent } from './useWrapEvent'

const isMovingUp = (e: FocusEvent, element: HTMLElement) => {
  const previousFocus = getNextFocusTarget(e)
  const relationship =
    previousFocus && element.compareDocumentPosition(previousFocus as Node)
  return relationship === Node.DOCUMENT_POSITION_FOLLOWING
}

export interface UseCollectionNavProps<E extends HTMLElement> {
  element: E | null
  onKeyDown?: (e: KeyboardEvent<E>) => void
  onFocus?: (e: FocusEvent<E>) => void
  onBlur?: (e: FocusEvent<E>) => void
}

export const useCollectionNav = <E extends HTMLElement = HTMLElement>({
  element,
  onBlur,
  onFocus,
  onKeyDown,
}: UseCollectionNavProps<E>) => {
  const [focusedItem, setFocusedItem] = useState<HTMLElement | null>(null)
  const [focusInside, setFocusInside] = useState(false)

  const handleArrowKey = (
    e: KeyboardEvent<E>,
    direction: number,
    initial: number
  ) => {
    const newFocusedItem = moveFocus(direction, initial, element)
    if (newFocusedItem) {
      e.preventDefault()
      setFocusedItem(newFocusedItem)
    }
  }

  const handleKeyDown = (e: KeyboardEvent<E>) => {
    switch (e.key) {
      case 'ArrowUp':
        e.preventDefault()
        handleArrowKey(e, -1, -1)
        break
      case 'ArrowDown':
        e.preventDefault()
        handleArrowKey(e, 1, 0)
        break
    }
  }

  const handleFocus = (e: FocusEvent<E>) => {
    setFocusInside(true)
    if (e.target === element) {
      if (focusedItem && element.contains(focusedItem)) {
        focusedItem.focus()
      } else {
        const tabStops = getTabStops(element)
        const toFocus = isMovingUp(e, element)
          ? tabStops[tabStops.length - 1]
          : tabStops[0]
        toFocus.focus()
      }
    }
  }

  const handleBlur = () => {
    setFocusInside(false)
  }

  return {
    onBlur: useWrapEvent(handleBlur, onBlur),
    onFocus: useWrapEvent(handleFocus, onFocus),
    onKeyDown: useWrapEvent(handleKeyDown, onKeyDown),
    tabIndex: focusInside ? undefined : 0,
  }
}
