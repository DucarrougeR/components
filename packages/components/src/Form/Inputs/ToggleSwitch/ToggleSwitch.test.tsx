import 'jest-styled-components'
import * as React from 'react'
import { mountWithTheme } from '../../../../../test/utils/create_with_theme'
import { assertSnapshot } from '../../../../../test/utils/snapshot'
import { ToggleSwitch } from './ToggleSwitch'

test('ToggleSwitch default', () => {
  assertSnapshot(<ToggleSwitch />)
})

test('ToggleSwitch on set to true', () => {
  assertSnapshot(<ToggleSwitch on={true} />)
})

test('ToggleSwitch on set to false', () => {
  assertSnapshot(<ToggleSwitch on={false} />)
})

test('ToggleSwitch that is disabled', () => {
  assertSnapshot(<ToggleSwitch on={true} disabled />)
})

test('Big ToggleSwitch', () => {
  assertSnapshot(<ToggleSwitch size={60} />)
})

test('Should trigger onChange handler', () => {
  let counter = 0
  const handleChange = () => counter++

  const wrapper = mountWithTheme(<ToggleSwitch onChange={handleChange} />)

  wrapper.find('input').simulate('change', { target: { value: '' } })
  expect(counter).toEqual(1)
})
