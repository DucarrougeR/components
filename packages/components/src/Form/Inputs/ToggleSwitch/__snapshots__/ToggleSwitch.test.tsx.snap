// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Big ToggleSwitch 1`] = `
.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  height: 100%;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 1;
}

.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  cursor: pointer;
  display: inline-block;
  height: 3.75rem;
  position: relative;
  vertical-align: middle;
  width: 6.5625rem;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border-radius: 3.75rem;
  bottom: 0;
  background-color: #C1C6CC;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.c5:hover {
  box-shadow: 0 0 .01rem 0.01rem rgba(100,81,138,0.5);
}

.c7 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border-radius: 50%;
  bottom: 0.375rem;
  background-color: #FFFFFF;
  height: 3rem;
  left: 0.375rem;
  position: absolute;
  width: 3rem;
}

.c6 {
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c4 {
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c2:focus + .c0 {
  box-shadow: 0 0 0 0.2rem rgba(100,81,138,0.4);
}

<div
  className="c0 c1"
  cursor="pointer"
>
  <input
    className="c2 c0 c3"
    role="switch"
    type="checkbox"
  />
  <div
    className="c4 c0 c5"
  >
    <div
      className="c6 c0 c7"
    />
  </div>
</div>
`;

exports[`ToggleSwitch default 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  cursor: pointer;
  display: inline-block;
  height: 1.25rem;
  position: relative;
  vertical-align: middle;
  width: 2.1875rem;
}

.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  height: 100%;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 1;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border-radius: 1.25rem;
  bottom: 0;
  background-color: #C1C6CC;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.c5:hover {
  box-shadow: 0 0 .01rem 0.01rem rgba(100,81,138,0.5);
}

.c7 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border-radius: 50%;
  bottom: 0.125rem;
  background-color: #FFFFFF;
  height: 1rem;
  left: 0.125rem;
  position: absolute;
  width: 1rem;
}

.c6 {
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c4 {
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c2:focus + .c0 {
  box-shadow: 0 0 0 0.2rem rgba(100,81,138,0.4);
}

<div
  className="c0 c1"
  cursor="pointer"
>
  <input
    className="c2 c0 c3"
    role="switch"
    type="checkbox"
  />
  <div
    className="c4 c0 c5"
  >
    <div
      className="c6 c0 c7"
    />
  </div>
</div>
`;

exports[`ToggleSwitch on set to false 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  cursor: pointer;
  display: inline-block;
  height: 1.25rem;
  position: relative;
  vertical-align: middle;
  width: 2.1875rem;
}

.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  height: 100%;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 1;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border-radius: 1.25rem;
  bottom: 0;
  background-color: #C1C6CC;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.c5:hover {
  box-shadow: 0 0 .01rem 0.01rem rgba(100,81,138,0.5);
}

.c7 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border-radius: 50%;
  bottom: 0.125rem;
  background-color: #FFFFFF;
  height: 1rem;
  left: 0.125rem;
  position: absolute;
  width: 1rem;
}

.c6 {
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c4 {
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c2:focus + .c0 {
  box-shadow: 0 0 0 0.2rem rgba(100,81,138,0.4);
}

<div
  className="c0 c1"
  cursor="pointer"
>
  <input
    aria-checked={false}
    checked={false}
    className="c2 c0 c3"
    role="switch"
    type="checkbox"
  />
  <div
    className="c4 c0 c5"
  >
    <div
      className="c6 c0 c7"
    />
  </div>
</div>
`;

exports[`ToggleSwitch on set to true 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  cursor: pointer;
  display: inline-block;
  height: 1.25rem;
  position: relative;
  vertical-align: middle;
  width: 2.1875rem;
}

.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  height: 100%;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 1;
}

.c7 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border-radius: 50%;
  bottom: 0.125rem;
  background-color: #FFFFFF;
  height: 1rem;
  left: 0.125rem;
  position: absolute;
  width: 1rem;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border-radius: 1.25rem;
  bottom: 0;
  background-color: #9785F2;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.c5:hover {
  box-shadow: 0 0 .01rem 0.01rem rgba(100,81,138,0.5);
}

.c6 {
  -webkit-transform: translateX(0.9375rem);
  -ms-transform: translateX(0.9375rem);
  transform: translateX(0.9375rem);
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c4 {
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c2:focus + .c0 {
  box-shadow: 0 0 0 0.2rem rgba(100,81,138,0.4);
}

<div
  className="c0 c1"
  cursor="pointer"
>
  <input
    aria-checked={true}
    checked={true}
    className="c2 c0 c3"
    role="switch"
    type="checkbox"
  />
  <div
    className="c4 c0 c5"
  >
    <div
      className="c6 c0 c7"
    />
  </div>
</div>
`;

exports[`ToggleSwitch that is disabled 1`] = `
.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  height: 100%;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 1;
}

.c7 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border-radius: 50%;
  bottom: 0.125rem;
  background-color: #FFFFFF;
  height: 1rem;
  left: 0.125rem;
  position: absolute;
  width: 1rem;
}

.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  display: inline-block;
  height: 1.25rem;
  position: relative;
  vertical-align: middle;
  width: 2.1875rem;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border-radius: 1.25rem;
  bottom: 0;
  background-color: #9785F2;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.c8 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border-radius: 1.25rem;
  bottom: 0;
  background-color: #C1C6CC;
  left: 0;
  opacity: 0.4;
  position: absolute;
  right: 0;
  top: 0;
}

.c6 {
  -webkit-transform: translateX(0.9375rem);
  -ms-transform: translateX(0.9375rem);
  transform: translateX(0.9375rem);
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c4 {
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c2:focus + .c0 {
  box-shadow: 0 0 0 0.2rem rgba(100,81,138,0.4);
}

<div
  className="c0 c1"
>
  <input
    aria-checked={true}
    checked={true}
    className="c2 c0 c3"
    disabled={true}
    role="switch"
    type="checkbox"
  />
  <div
    className="c4 c0 c5"
  >
    <div
      className="c6 c0 c7"
    />
  </div>
  <div
    className="c0 c8"
  />
</div>
`;
